src/Datadog.Trace/Agent/MessagePack/CachedSerializer.cs

```diff
@@ -35,6 +43,17 @@ public async Task SerializeAsync<T>(Stream stream, T obj, IFormatterResolver res
 
                 int length = MessagePackSerializer.Serialize(ref buffer, 0, obj, resolver);
 
+#if DEBUG
+                if (_writePayloadToFile && length > 3)
+                {
+                    // If there are actual traces present, then write
+                    // Trim down to the actual length to write to file
+                    var trimmed = new byte[length];
+                    System.Array.Copy(buffer, 0, trimmed, 0, length);
+                    File.WriteAllBytes(Path.Combine(_payloadFilePath, $"payload-{System.DateTime.Now.Ticks}.txt"), trimmed);
```

Comment:
NIT: I know this is only for debugging purposes, but this can be optimized for no allocation, no copies and not blocking.

Reply:
Removed for now